Diggy automatic restart

Jayefuu opened the pull request:
- Added a discord role to automatically ping diggy users
- Added role to role reactions on carlbot
- Added diggy specific restart command
- Win condition is a certain level of mining productivity. Level to be decided.
- Would be nice if the win condition was randomised each time but I need help to get that to work.
- Would be nice if the win condition was decided on the previous game and passed into the next one so that the discord post could say what the next map win condition will be. But I don't have a clue how to approach that.

Jayefuu committed:
Diggy automatic restart

grilledham commented on map_gen/maps/diggy/modules/restart_command.lua:
```diff
@@ -0,0 +1,51 @@
+        --    return true
+        --end
+        if game.player.force.technologies["mining-productivity-4"].level < restart_difficulty  then
+            player.print("Mine harder dwarf! You must reach Mining Productivity "..restart_difficulty.." to start a new mining venture!")
```
Suggestion: Use localised string.

grilledham commented on map_gen/maps/diggy/modules/restart_command.lua:
```diff
@@ -0,0 +1,51 @@
+
+return function(config)
+    local map_promotion_channel = Discord.channel_names.map_promotion
+    local danger_ore_role_mention = Discord.role_mentions.diggy
```
```suggestion
    local diggy_role_mention = Discord.role_mentions.diggy
```

Jayefuu replied:
Fixed in latest commit.

grilledham commented on map_gen/maps/diggy/modules/restart_command.lua:
```diff
@@ -0,0 +1,51 @@
+    Restart.set_start_game_data({type = Restart.game_types.scenario, name = config.scenario_name or 'diggy-next'})
+
+    local function can_restart(player)
+        --if player.admin then
+        --    return true
+        --end
```
Why no admin restarts?

Jayefuu replied:
Disabled for testing. Latest commit adds them back in along with some feedback for the admin as to whether it reached victory conditions.

grilledham commented on map_gen/maps/diggy/scenario.lua:
```diff
@@ -3,9 +3,12 @@ local Config = require 'map_gen.maps.diggy.config'
 local ScenarioInfo = require 'features.gui.info'
 local RS = require 'map_gen.shared.redmew_surface'
 local Event = require 'utils.event'
-local type = type
```
Why remove, it is used?

Jayefuu replied:
It wasn't used and gave me an error. I assumed that VS code added it automatically as it sometimes likes to do.

grilledham replied:
It is used on line 30 for example.

grilledham commented on resources/discord.lua:
```diff
@@ -14,6 +14,7 @@ return {
         danger_ore = '<@&793231011144007730>',
-        moderator = '<@&454192594633883658>'
+        moderator = '<@&454192594633883658>',
+        diggy = '@921476458076061718'
```
Does this not need the `<@&` like the other roles?

Jayefuu replied:
Thanks I missed that. Will be fixed in next commit.

grilledham reviewed (commented):
> Would be nice if the win condition was decided on the previous game and passed into the next one so that the discord post could say what the next map win condition will be. But I don't have a clue how to approach that.

That could probably be done with scenario data. Sounds like an optional extra that could be added later.

This mostly looks good. I haven't tested it though, have you?

Jayefuu committed:
Corrections and ease of use

- Moved restart_command from modules to feature folder
- Added some admin messages as usually it's not clear if the victory condition was reached. Allows admin to make a better choice when using /abort.
- Added randomisation of victory condition. Ranges from lvl 5 (5k science) to lvl 15 (30k science)
- Adjusted the printed level to align with in-game research.

Jayefuu committed:
Updates as per request

- added localisation for restart message
- fixed discord tag formatting
- added some colour feedback for admins to make it easier to see the message among the restart messages.

Jayefuu committed:
Fixed discord role formatting. Woops

Jayefuu commented:
Thanks for the review @grilledham.

I think I've corrected everything. I tested it on S10 with the live role just once and it all seems to be working.

grilledham commented on map_gen/maps/diggy/feature/restart_command.lua:
```diff
@@ -0,0 +1,67 @@
+    local function can_restart(player)
+        local adjusted_difficulty = restart_difficulty - 1 --because the labels in game are different from the technologies[].level. This adjusts the printed messages so it's correct for player.
+        if player.admin then
+            if game.player.force.technologies["mining-productivity-4"].level < restart_difficulty  then
```
```suggestion
            if game.forces.player.technologies["mining-productivity-4"].level < restart_difficulty  then
```
Else it will error when run from the server console.
Same again below on line 39

grilledham commented on map_gen/maps/diggy/feature/restart_command.lua:
```diff
@@ -0,0 +1,67 @@
+            return true
+        end
+        if game.player.force.technologies["mining-productivity-4"].level < restart_difficulty  then
+            game.print({'command_description.diggy_restart_condition_not_met', adjusted_difficulty},Color.fail)
```
```suggestion
            player.print({'command_description.diggy_restart_condition_not_met', adjusted_difficulty},Color.fail)
```

Jayefuu committed:
Updated as per suggestions

Jayefuu commented:
@grilledham I've corrected the remaining 3 mistakes mentioned above.

grilledham commented on map_gen/maps/diggy/scenario.lua:
```diff
@@ -3,8 +3,11 @@ local Config = require 'map_gen.maps.diggy.config'
 local Event = require 'utils.event'
-local type = type
 local pairs = pairs
+type = type
```
```suggestion
local type = type
```

Jayefuu committed:
Update map_gen/maps/diggy/scenario.lua

Co-authored-by: grilledham <jrdgillham@gmail.com>
